### ISSUE TITLE
Evaluation of self-trained mpt-7b-instruct for mosaic checkpoint broken

### ISSUE THREAD

**Issue by olaf-beh** [mosaicml/llm-foundry#369]
## Environment

```
PyTorch version: 2.0.1+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Debian GNU/Linux 11 (bullseye) (x86_64)
GCC version: (Debian 10.2.1-6) 10.2.1 20210110
Clang version: 11.0.1-2
CMake version: version 3.25.0
Libc version: glibc-2.31

Python version: 3.9.2 (default, Feb 28 2021, 17:03:44)  [GCC 10.2.1 20210110] (64-bit runtime)
Python platform: Linux-5.16.0-0.bpo.4-amd64-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA A100-SXM4-80GB

Nvidia driver version: 520.61.05
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   48 bits physical, 48 bits virtual

Versions of relevant libraries:
[pip3] numpy==1.24.3
[pip3] pytorch-ranger==0.1.1
[pip3] torch==2.0.1+cu118
[pip3] torch-optimizer==0.3.0
[pip3] torchdata==0.6.1
[pip3] torchmetrics==0.11.3
[pip3] torchtext==0.15.2
[pip3] torchvision==0.15.2
[pip3] triton==2.0.0
[pip3] triton-pre-mlir==2.0.0
[conda] Could not collect
```
## To reproduce

1. Install llm-foundary
2. `cd llm-foundry/scripts`
3. Run evaluation on instruction tuned (dolly_hhrlhf) mpt-7b model mosaic checkpoint:

```bash
(llmfoundry-venv) $ python eval/eval.py eval/yamls/mpt7b_eval.yaml icl_tasks=eval/yamls/winograd.yaml \
load_path=mpt7b_dolly_sft/checkpoints/ep2-ba2000-rank0.pt
...
RuntimeError: Error(s) in loading state_dict for ComposerMPTCausalLM:
        Missing key(s) in state_dict: "model._fsdp_wrapped_module.transformer.wpe._fsdp_wrapped_module.weight",
...
        size mismatch for model._fsdp_wrapped_module.transformer.wte._fsdp_wrapped_module.weight: \
        copying a param with shape torch.Size([50432, 4096]) from checkpoint, the shape in current model is \
        torch.Size([50368, 4096]).
...
RuntimeError: Failed to load checkpoint due to missing or unexpected keys in state_dict. \
This is likely due to a change in the model architecture. If this is intentional, you can set \
load_strict_model_weights=False in the Trainer.
```

## Expected behavior

Working evaluation for mpt-7b-instruct checkpoint, in particular

* No RuntimeError
* no vocabulary size mismatch.

### TL;DR

I finally got it running by doing 

1. Change `vocab_size 50368` to `vocab_size 50432` in eval yaml
2. Adding `load_strict_model_weights=False,` to `Trainer` in `eval.py`

But even if I found a a workaround, following issues should be addressed imho:

* All training and eval yamls in llm-foundary repository got the wrong vocabulary dimension of 50368 (see details below)?
* I am not sure if setting `load_strict_model_weights=False` might have some unexpected sideeffects on evaluation?

### Analysis

Convert checkpoint to HF format works:

```bash
python inference/convert_composer_to_hf.py --composer_path mpt7b_dolly_sft/checkpoints/ep2-ba2000-rank0.pt \
--hf_output_path mpt7b_dolly_sft_2ep_hf --output_precision bf16
...
HF checkpoint folder successfully created at mpt7b_dolly_sft_2ep_hf.
Done.
```
Apparently HF converted model is using `vocab_size 50432`!:

```bash
grep vocab_size mpt7b_dolly_sft_2ep_hf/config.json 
  "vocab_size": 50432
```

Searching for vocabulary size I find following inconsitent information:

* In all mosaic training Yamls, e.g. `pretrain/mpt-7b.yaml`:
  > `model.vocab_size: 50368`
* [HF mosaicml/mpt-7b](https://huggingface.co/mosaicml/mpt-7b): 
  > `vocab size 	50432`
* From [mpt-7b blog post]9https://www.mosaicml.com/blog/mpt-7b):
  > The tokenizer has a vocabulary size of 50257, but we set the model vocabulary size to 50432.

Changing `vocab_size 50368` to `vocab_size 50432` in `mpt7b_eval.yaml`:

```bash
python eval/eval.py eval/yamls/mpt7b_eval.yaml icl_tasks=eval/yamls/winograd.yaml \
load_path=mpt7b_dolly_sft/checkpoints/ep2-ba2000-rank0.pt
...
RuntimeError: Error(s) in loading state_dict for ComposerMPTCausalLM:
        Missing key(s) in state_dict: "model._fsdp_wrapped_module.transformer.wpe._fsdp_wrapped_module.weight", 
...
RuntimeError: Failed to load checkpoint due to missing or unexpected keys in state_dict. This is likely due to a change in the model architecture. If this is 
intentional, you can set load_strict_model_weights=False in the Trainer.
```
Adding `load_strict_model_weights=False,` to `eval.py` works finally:

```bash
python eval/eval.py eval/yamls/mpt7b_eval.yaml icl_tasks=eval/yamls/winograd.yaml \
load_path=mpt7b_dolly_sft/checkpoints/ep2-ba2000-rank0.pt
...
Found these missing keys in the checkpoint: model._fsdp_wrapped_module.transformer.wpe._fsdp_wrapped_module.weight, \
model._fsdp_wrapped_module.transformer.blocks.0._fsdp_wrapped_module.norm_1.bias,
...
model._fsdp_wrapped_module.transformer.blocks.31._fsdp_wrapped_module.ffn.down_proj.bias, \
model._fsdp_wrapped_module.transformer.norm_f._fsdp_wrapped_module.bias
...
Ran eval in: 67.24968338012695 seconds
metrics/winograd/0-shot/InContextLearningMultipleChoiceAccuracy: 0.831501841545105
```
The winograd accuracy of 0.831 is consistent with reported pre-trained only 0.878 (fine-tuning of bmp-7b on dolly_hhrlhf results in worse average benchmark results as documented by [RedPajama](https://www.together.xyz/blog/redpajama-7b) on HELM).

**Comment by olaf-beh** [comment]
My `mpt7b_eval.yaml` based on `mpt_eval.yaml` after editing

* Using `model` from `finetune/7b_dolly_sft.yaml` instead of mpt-125m model parameters
* `precision: fp32` -> `precision: amp_bf16`
* `vocab_size: 50368` -> `vocab_size: 50432`

```yaml
max_seq_len: 2048
tokenizer_name: EleutherAI/gpt-neox-20b
seed: 1
# precision: fp32
precision: amp_bf16

# Tokenizer
tokenizer:
  name: ${tokenizer_name}
  kwargs:
    model_max_length: ${max_seq_len}

# mpt-7b taken from train/yamls/pretrain/mpt-7b.yaml
model:
  name: mpt_causal_lm
  init_device: meta
  d_model: 4096
  n_heads: 32
  n_layers: 32
  expansion_ratio: 4
  max_seq_len: ${max_seq_len}
  vocab_size: 50432
  # vocab_size: 50368
  attn_config:
    attn_impl: triton

load_path: # Add your (optional) Composer checkpoint path here!

device_eval_batch_size: 16

# FSDP config for model sharding
fsdp_config:
  sharding_strategy: FULL_SHARD
  mixed_precision: FULL
# ...
```

**Comment by olaf-beh** [comment]
I accidently started `eval.py` with `python` instead of `composer`. But I still get the same Error(s) as above and I have to change `vocab_size 50368` to `vocab_size 50432` in eval yaml file and add `load_strict_model_weights=False` to Trainer in `eval.py` as with using `python` start command inorder to run evaluation successfully.

**Comment by alextrott16** [comment]
Thank you for catching this. The `vocab_size: 50368` setting is an error on our end. It should be 50432. We'll fix that.

@eracah Can you lend any insight into why `load_strict_model_weights=False` is required? It might come down to using a sharded checkpoint versus a monolithic checkpoint. But I'm not sure.

**Comment by dakinggg** [comment]
Hi @olaf-beh, if you're loading a self trained checkpoint, you will need to make sure that the model configuration is _exactly_ the same for eval as it was for training. I suspect you changed something by mistake, and that is why you needed to `load_strict_model_weights=False`. Does that help?

**Comment by dakinggg** [comment]
Closing due to inactivity. Please feel free to open a new issue if you are still encountering problems.
